apps/api/src/app/events/usecases/trigger-event/message-matcher.service.ts

```diff
@@ -80,61 +102,74 @@ export class MessageMatcher {
   private async handleGroupFilters(
     filter: StepFilter,
     variables: IFilterVariables,
-    command: SendMessageCommand
+    command: SendMessageCommand,
+    filterProcessingDetails: FilterProcessingDetails
   ): Promise<boolean> {
     if (filter.value === 'OR') {
-      return await this.handleOrFilters(filter, variables, command);
+      return await this.handleOrFilters(filter, variables, command, filterProcessingDetails);
     }
 
     if (filter.value === 'AND') {
-      return await this.handleAndFilters(filter, variables, command);
+      return await this.handleAndFilters(filter, variables, command, filterProcessingDetails);
     }
 
     return false;
   }
 
-  private splitToSyncAsync(filter: StepFilter) {
-    const asyncFilters = findFilterPartByType(filter.children, 'webhook') as IWebhookFilterPart[];
-    const syncFilters = findFilterPartByType(filter.children, ['payload', 'subscriber']) as IFieldFilterPart[];
+  private splitFilters(filter: StepFilter) {
+    const webhookFilters = filter.children.filter((childFilter) => childFilter.on === 'webhook');
 
-    return { asyncFilters, syncFilters };
+    const otherFilters = filter.children.filter((childFilter) => childFilter.on !== 'webhook');
+
+    return { webhookFilters, otherFilters };
   }
 
   private async handleAndFilters(
     filter: StepFilter,
     variables: IFilterVariables,
-    command: SendMessageCommand
+    command: SendMessageCommand,
+    filterProcessingDetails: FilterProcessingDetails
   ): Promise<boolean> {
-    const { asyncFilters, syncFilters } = this.splitToSyncAsync(filter);
+    const { webhookFilters, otherFilters } = this.splitFilters(filter);
 
-    const foundSyncFilterMatches = syncFilters.filter((i) => this.processFilterEquality(variables, i));
-    if (syncFilters.length !== foundSyncFilterMatches.length) {
+    const matchedOtherFilters = await filterAsync(otherFilters, (i) =>
+      this.processFilter(variables, i, command, filterProcessingDetails)
+    );
```

Comment:
instead of calling the `processFilterEquality` specific function we do call now the `processFilter` that have the logic that decides what to do based on the filter `on` field